cms/lib/xblock/upstream_sync.py

```diff
@@ -0,0 +1,451 @@
+        except XBlockNotFoundError as exc:
+            raise BadUpstream(_("Linked library item was not found in the system")) from exc
+        return cls(
+            upstream_ref=downstream.upstream,
+            version_synced=downstream.upstream_version,
+            version_available=(lib_meta.draft_version_num if lib_meta else None),
+            # TODO: Previous line is wrong. It should use the published version instead, but the
+            # LearningCoreXBlockRuntime APIs do not yet support published content yet.
+            # Will be fixed in a follow-up task: https://github.com/openedx/edx-platform/issues/35582
+            # version_available=(lib_meta.published_version_num if lib_meta else None),
+            version_declined=downstream.upstream_version_declined,
+            error_message=None,
+        )
+
+
+def sync_from_upstream(downstream: XBlock, user: AbstractUser) -> None:
+    """
+    Update `downstream` with content+settings from the latest available version of its linked upstream content.
+
+    Preserves overrides to customizable fields; overwrites overrides to other fields.
+    Does not save `downstream` to the store. That is left up to the caller.
+
+    If `downstream` lacks a valid+supported upstream link, this raises an UpstreamLinkException.
+    """
+    link, upstream = _load_upstream_link_and_block(downstream, user)
+    _update_customizable_fields(upstream=upstream, downstream=downstream, only_fetch=False)
+    _update_non_customizable_fields(upstream=upstream, downstream=downstream)
+    downstream.upstream_version = link.version_available
+
+
+def fetch_customizable_fields(*, downstream: XBlock, user: AbstractUser, upstream: XBlock | None = None) -> None:
+    """
+    Fetch upstream-defined value of customizable fields and save them on the downstream.
+
+    If `upstream` is provided, use that block as the upstream.
+    Otherwise, load the block specified by  `downstream.upstream`, which may raise an UpstreamLinkException.
+    """
+    if not upstream:
+        _link, upstream = _load_upstream_link_and_block(downstream, user)
+    _update_customizable_fields(upstream=upstream, downstream=downstream, only_fetch=True)
+
+
+def _load_upstream_link_and_block(downstream: XBlock, user: AbstractUser) -> tuple[UpstreamLink, XBlock]:
+    """
+    Load the upstream metadata and content for a downstream block.
+
+    Assumes that the upstream content is an XBlock in an LC-backed content libraries. This assumption may need to be
+    relaxed in the future (see module docstring).
+
+    If `downstream` lacks a valid+supported upstream link, this raises an UpstreamLinkException.
+    """
+    link = UpstreamLink.get_for_block(downstream)  # can raise UpstreamLinkException
+    # We import load_block here b/c UpstreamSyncMixin is used by cms/envs, which loads before the djangoapps are ready.
+    from openedx.core.djangoapps.xblock.api import load_block  # pylint: disable=wrong-import-order
```

Comment:
It could be nicer in the future to move `UpstreamSyncMixin` to its own file? It seems fairly separable from everything else in this file.

Reply:
Yeah, I'm also thinking that might be necessary if we're to make these fields available in the LMS (for analytics) without exposing the whole upstream_sync API to LMS.

I'll leave it for now, but I will likely do this refactoring when it becomes necessary.